osv2/index_create.go

```diff
@@ -0,0 +1,161 @@
+	// converting from the domain model to the V2 model.
+	var vrs opensearchtools.ValidationResults
+
+	return CreateIndexRequest{
+		Index:               req.Index,
+		DocBody:             req.DocBody,
+		MasterTimeout:       req.MasterTimeout,
+		Timeout:             req.Timeout,
+		WaitForActiveShards: req.WaitForActiveShards,
+	}, vrs
+}
+
+// Validate validates the given CreateIndexRequest
+func (c *CreateIndexRequest) Validate() opensearchtools.ValidationResults {
+	var validationResults opensearchtools.ValidationResults
+
+	if c.Index == "" {
+		validationResults.Add(opensearchtools.NewValidationResult("Index not set at the CreateIndexRequest", true))
+	}
+
+	return validationResults
+}
+
+// NewCreateIndexRequest instantiates an CreateIndexRequest with default values
+func NewCreateIndexRequest() *CreateIndexRequest {
+	return &CreateIndexRequest{
+		MasterTimeout:       30 * time.Second,
+		Timeout:             30 * time.Second,
+		WaitForActiveShards: "1",
+	}
+}
+
+// WithIndex adds the index for CreateIndexRequest
+func (c *CreateIndexRequest) WithIndex(index string) *CreateIndexRequest {
+	c.Index = index
+	return c
+}
+
+// WithDocBody adds the body for CreateIndexRequest that contains detailed index setting
+func (c *CreateIndexRequest) WithDocBody(body io.Reader) *CreateIndexRequest {
+	c.DocBody = body
+	return c
+}
+
+// WithMasterTimeout adds the master timeout for CreateIndexRequest
+// it defines how long to wait for a connection to the master node. Default is 30s.
+func (c *CreateIndexRequest) WithMasterTimeout(d time.Duration) *CreateIndexRequest {
+	c.MasterTimeout = d
+	return c
+}
+
+// WithTimeout adds the timeout for CreateIndexRequest, it defines how long to wait for the request to return. Default is 30s
+func (c *CreateIndexRequest) WithTimeout(d time.Duration) *CreateIndexRequest {
+	c.Timeout = d
+	return c
+}
+
+// WithWaitForActiveShards to add the active shard options with CreateIndexRequest,
+// it specifies the number of active shards that must be available before OpenSearch processes the request. Default is 1
+func (c *CreateIndexRequest) WithWaitForActiveShards(s string) *CreateIndexRequest {
+	c.WaitForActiveShards = s
+	return c
+}
+
+// Do executes the [CreateIndexRequest] using the provided opensearch.Client.
+// If the request is executed successfully, then a [CreateIndexResponse] will be returned.
+// An error can be returned if
+//
+//   - Index is missing
+//   - The call to OpenSearch fails
+//   - The result json cannot be unmarshalled
+func (c *CreateIndexRequest) Do(ctx context.Context, client *opensearch.Client) (*opensearchtools.OpenSearchResponse[CreateIndexResponse], error) {
+	vrs := c.Validate()
+	if vrs.IsFatal() {
+		return nil, opensearchtools.NewValidationError(vrs)
+	}
+
+	osResp, rErr := opensearchapi.IndicesCreateRequest{
+		Body:                c.DocBody, // todo: are we sure about this? what will happen to nil?
```

Comment:
If `nil` is actually not valid that should be caught in `Validate()` and returned above.

Reply:
So in the end nil is valid. But in case there's no template matches the index name we will get the error.